### ISSUE TITLE
NullReferenceException for collection BeEquivalentTo

### ISSUE THREAD

**Issue by snekbaev** [fluentassertions/fluentassertions#917]
Hi,

following works with 5.4.1:
```c#
duplicatedProductVariants.Should().BeEquivalentTo(new[] { productVariant1Expectation, productVariant2Expectation },
    options => options.DateTimeCloseToUtcNow(precision)
        .NoEmptyGuid()
    .Using<string>(ctx => ctx.Subject.Should().NotBeNull()
        .And.NotBe(productVariant1Expectation.MediaFilename)
        .And.NotBe(productVariant2Expectation.MediaFilename))
    .When(info => info.SelectedMemberInfo.Name == nameof(ProductVariant.MediaFilename))
    .Excluding(x => x.NotesInstructions)
    .Excluding(x => x.ValidationErrors)
    .Excluding(x => x.IsValid));
```

however, fails with 5.4.2:
```
System.NullReferenceException : Object reference not set to an instance of an object.
   at FluentAssertions.Equivalency.GenericEnumerableEquivalencyStep.Handle(IEquivalencyValidationContext context, IEquivalencyValidator parent, IEquivalencyAssertionOptions config)
   at FluentAssertions.Equivalency.EquivalencyValidator.AssertEqualityUsing(IEquivalencyValidationContext context)
   at FluentAssertions.Equivalency.EquivalencyValidator.AssertEquality(EquivalencyValidationContext context)
   at FluentAssertions.Collections.CollectionAssertions`2.BeEquivalentTo[TExpectation](IEnumerable`1 expectation, Func`2 config, String because, Object[] becauseArgs)
   at REDACTED
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at NUnit.Framework.Internal.AsyncInvocationRegion.AsyncTaskInvocationRegion.WaitForPendingOperationsToComplete(Object invocationResult) in C:\src\nunit\nunit\src\NUnitFramework\framework\Internal\AsyncInvocationRegion.cs:line 113
   at NUnit.Framework.Internal.Commands.TestMethodCommand.RunAsyncTestMethod(TestExecutionContext context) in C:\src\nunit\nunit\src\NUnitFramework\framework\Internal\Commands\TestMethodCommand.cs:line 96
```

Might be related to other open issues related to collections comparisons.

Thank you!

**Comment by jnyrup** [comment]
Thanks for opening this issue.

It seems you're using several custom constructs, e.g. `DateTimeCloseToUtcNow` and `NoEmptyGuid`.
Can you provide a [complete, minimal and verifiable example](https://stackoverflow.com/help/mcve)?
That would help tracing down the issue.

**Comment by krajek** [comment]
@snekbaev I tried to take a look at the code but the method (in fact whole class) was not changed between `5.4.1` and `5.4.2`. I am willing to help you but we need more details.

**Comment by snekbaev** [comment]
@krajek @jnyrup sorry for the delay :)

Noticed version 5.5 which also has failed for me. Actually more failing tests, I think there are some breaking changes with Func<Task> act .Should().Throw handling. Seems like it is not waiting. Will check those after main issue is resolved.

Here is a snippet I made in LinqPad:

```
void Main()
{
	var pv1 = new Test {MediaFilename = "1"};
	var pv2 = new Test {MediaFilename = "2"};

	new[] { pv1 }.Should().BeEquivalentTo
		( new[] { pv2 },
		 options => options.Using<string>
							(ctx => ctx.Subject.Should().NotBeNull().And.NotBe(pv2.MediaFilename))
						   .When(info => info.SelectedMemberInfo.Name == nameof(Test.MediaFilename))
						   );
}

public class Test
{
	public string MediaFilename { get; set; }
}
```
this passes with 5.4.1 and fails with 5.4.1+.

Thank you!

**Comment by jnyrup** [comment]
Thanks for the example, I can now reproduce your problem.

From the XML summary of `SelectedMemberInfo`
> Gets the `SelectedMemberInfo` of the member that returned the current object, or `null` if the current object represents the root object.

When `BeEquivalent` applies then `When` on the root, `SelectedMemberInfo` is `null` and `.Name` throws an NRE.

I can confirm that this behavior changed from 5.4.1 to 5.4.2 and that is it due to #886.
@dennisdoomen can you comment on whether that change in behavior is intended?

There are two immediate ways to solve this:

check that `SelectedMemberInfo` is not `null`
```c#
.When(info => info.SelectedMemberInfo != null && info.SelectedMemberInfo.Name == nameof(Test.MediaFilename)));
```

or use `SelectedMemberPath` which should never be `null`

```c#
.When(info => info.SelectedMemberPath.EndsWith(nameof(Test.MediaFilename))));
```

**Comment by dennisdoomen** [comment]
Hmm, I need to check the example. I do know that #886 fixed the problem that the custom rule was not even executed. So this potentially new bug may have already existed.

**Comment by dennisdoomen** [comment]
I can confirm that `SelectMemberInfo` is supposed to return `null` if the current object is representing the root object. See also the comments on the property [here](https://github.com/fluentassertions/fluentassertions/blob/7738a652886af2e2ca0918ce71b197f7e06d8441/Src/FluentAssertions/Equivalency/IMemberInfo.cs#L14). The problem was that the `When` block was never called. This was fixed in #886.
